Task/finishaccountbind

thommey opened the pull request:
- Central setaccount() function to update account and call binds
- Normalized m->account values (empty string = unknown, "*" = not logged in, otherwise accountname)
- raw and rawt accept wildcards
- moved most of account functionality to irc.mod because it's only relevant in the context of channels
- improved .status reporting for account tracking
- request account-notify and extended-join by default

vanosg commented on src/mod/irc.mod/chan.c:
```diff
@@ -212,6 +212,33 @@ static void update_idle(char *chname, char *nick)
+/* set user account on all members on all channels,
+ * trigger account bind if account state was not "unknown" (empty string)
+ */
+// TODO should be in irc.mod
```
It is!

vanosg commented on src/mod/irc.mod/chan.c:
```diff
@@ -212,6 +212,33 @@ static void update_idle(char *chname, char *nick)
+        /* account was known */
+        if (m->account[0]) {
+          if (!strcmp(account, "*")) {
+            putlog(LOG_JOIN, chan->dname, "%s!%s has logged out of their account", nick, m->userhost);
+          } else {
+            putlog(LOG_JOIN, chan->dname, "%s!%s logged in to their account %s", nick, m->userhost, account);
```
I can see LOG_JOIN makes sense particularly when a user is joining a channel with this information, but an idle user could be present in a channel for a week and then authenticate, so I would tend to view an account as a server-related activity moreso than a join activity. My first instinct was to go either DEBUG or SERVER... but to be fair, 'aways' are listed under join. Hmm hmm...

And separately, pure aesthetics, do we want to log the entire usermask? Or just the nick

thommey replied:
gotaway logs exactly the same way including the full hostmask, would be a discussion about that as well.
Channel related I believe we only have JOIN and MODES but those changes are more subtle than that, I'll leave this for now but make a mental note to discuss both before release (look at all LOG_JOIN uses and stuff)

vanosg commented on src/mod/irc.mod/chan.c:
```diff
@@ -1216,25 +1241,11 @@ static int got352or4(struct chanset_t *chan, char *user, char *host,
   /* Update accountname in channel records, 0 means logged out */
   /* A 0 is not a change from "" */
   if (account) {
```
This gets tricky.... this function is triggered for every WHOX, but the account field we're looking for depends on the WHOX format field being "c%%chnufat,222". If a Tcl script sends "%cuhnar,42" as the format field, Eggdrop does this:
```
[22:04:05] BeerBot!user@b0rg-C0E59E12.foo.net logged in to their account :realname
```
And .channel shows the account name as ":realname"

thommey replied:
we ignore the response in got354 if the query type isn't 222 so mismatch isn't possible unless scripts are malicious, the colon must be from gotaccount

vanosg replied:
nuh-uh!

vanosg commented on eggdrop.conf:
```diff
@@ -1158,17 +1158,21 @@ server add ssl.example.net +7000
+# To request the message-tags feature via CAP, set this to 1
+set message-tags 1
+
+# To request the account-tag feature via CAP, set this to 1
+# This should only be enabled if necessary for imperfect account tracking
+# see doc/ACCOUNT for details
```
A user may have other reasons they want to enable this than account-tracking, we may want to soften this language

vanosg commented on eggdrop.conf:
```diff
@@ -1158,17 +1158,21 @@ server add ssl.example.net +7000
 
 # To request the account-notify feature via CAP, set this to 1
-#set account-notify 0
+set account-notify 1
 
 # To request the extended-join feature via CAP, set this to 1
-#set extended-join 0
+set extended-join 1
 
 # To request the invite-notify feature via CAP, set this to 1
 #set invite-notify 0
```
I was going to say I was wary of enabling this by default, but I talked myself out of it, this is a good idea. But that inner-monologue left me thinking we need to add a big warning above this section saying "just because you enable it here doesn't mean the server supports it" or some variation. I definitely see people thinking these are enabled simply because it is set here.

thommey replied:
I think your warning on top of the CAP block in the configuration should be enough

vanosg commented on eggdrop.conf:
```diff
@@ -1158,17 +1158,21 @@ server add ssl.example.net +7000
-# To request the message-tags feature via CAP, set this to 1. NOTE: Enabling
-# this feature may interfere with RAW binds in scripts.
-#set message-tags 0
+# To request the message-tags feature via CAP, set this to 1
```
Does this no really longer affect RAW binds after these changes? IIRC, a rawt bind will stop, and the raw bind table will not be checked?

Also, do we really want this enabled by default in a general sense? I suspect you did it for account-tag, but ... enh. Change.

thommey replied:
can you find out if we need it or if it works separately from account-tag? I disabled it for now and only set account-tag

vanosg replied:
account-tag will insert a tag to a message, even if message-tags is not enabled.

vanosg reviewed (commented):
Awesome!

vanosg commented on src/mod/irc.mod/irc.c:
```diff
@@ -1045,13 +1061,59 @@ static void flush_modes()
+  }
+
+  if (whox && notify && extjoin) {
+    dprintf(idx, "%s", "    Account tracking: Perfect\n");
```
Can we go back to "enabled" or some other word here? Supported? Fully enabled? I know we've been using this word to describe the scenario, but I think the intended meaning without additional context would be lost on a novice user.

thommey replied:
perfect idea

vanosg commented on src/mod/irc.mod/irc.c:
```diff
@@ -1045,13 +1061,59 @@ static void flush_modes()
+      if (tag) {
+        dprintf(idx, "%s", "      - account-tag present    => Accounts will update whenever someone messages a channel or this bot\n");
+      }
+      dprintf(idx, "%s", "      See doc/ACCOUNT for more details");
```
Need a \n

vanosg commented on src/mod/irc.mod/irc.c:
```diff
@@ -1045,13 +1061,59 @@ static void flush_modes()
+                    details ? "" : ", see .status all for details");
+    } else {
+      dprintf(idx, "    Account tracking: Best-effort\n");
+      if (!whox) {
+        dprintf(idx, "%s", "      - WHOX missing           => Accounts will not be known after Eggdrop joins a channel\n");
+      }
+      if (!notify) {
+        dprintf(idx, "%s", "      - account-notify missing => Accounts will not update immediately when users log in or out\n");
+      }
+      if (!extjoin) {
+        dprintf(idx, "%s", "      - extended-join missing  => Accounts will not be known immediately when a user joins\n");
+      }
+      if (tag) {
+        dprintf(idx, "%s", "      - account-tag present    => Accounts will update whenever someone messages a channel or this bot\n");
```
I like this idea... maybe add the inverse comment as well, what *does* work? 
```
      if (!whox) {
        dprintf(idx, "%s", "      - WHOX missing           => Accounts will not be known after Eggdrop joins a channel\n");
      } else {
        dprintf(idx, "%s", "      - WHOX present           => Accounts will update when Eggdrop joins a channel\n");
      }
```
to give the user a full idea of what it does/doesn't do?

thommey commented:
ready for review again I think

thommey commented:
docs need to be generated after merge

thommey committed:
Move account bind to irc.mod and normalize account bind.

accountname being "" means unknown, accountname being "*" is not logged into services.
Account bind is triggered whenever a known account changes.

Patch by: thommey

thommey committed:
Raw and rawt bind now support wildcards, necessary for account-tag usage

thommey committed:
Document how to wildcard match against an asterisk (*)

thommey committed:
Eggdrop no longer interprets WHO response it didn't request in the WHOX case

The response could have different arguments, so we now validate our query type
parameter is in the response. Tcl scripts wanting to poll WHO and eggdrop
to parse it, must send query type 222 in the same format as eggdrop requests it.

Found by: Geo
Patch by: thommey

thommey committed:
Offer the option to use account-tag for imperfect tracking

It must be explicitely enabled in the configuration. The requirements for
perfect tracking are now enabled by default.

Patch by: thommey

thommey committed:
Format fix for .channel

thommey committed:
Move away and account change logs from console +j to +m

Found by: Geo
Patch by: Geo, thommey

vanosg committed:
Add ACCOUNTS doc

Patch by: Geo

thommey commented on doc/sphinx_source/using/accounts.rst:
```diff
@@ -0,0 +1,81 @@
+
+Best-Effort Account Tracking
+----------------------------
+If a server only supports some, but not all, of the required capabilities, Eggdrop will switch to 'best effort' account tracking. This means Eggdrop will update account statuses whenever it sees account information, but Eggdrop cannnot guarantee the accuracy of account association at any given time.
```
"Eggdrop cannnot guarantee the accuracy of account association at any given time" I would say "Eggdrop cannot guarantee that the account association is up to date"

thommey commented on doc/sphinx_source/using/accounts.rst:
```diff
@@ -0,0 +1,81 @@
+
+If a server does not support WHOX, Eggdrop will not be able to determine the accounts associated with users already on a channel before Eggdrop joined. There is no reliable way to update this information.
+
+One workaround to significantly increase the accuracy of account tracking for scripts in a 'best effort' scenario would be to issue a WHOX query (assuming the server supports it), wait for the reply from the eserver, and then query for the account information.
```
typo "eserver" and I would mention an explicit code example to do a minutely /who on all channels:
```tcl
bind cron - * updatewho
proc updatewho {args} {
  foreach chan [channels] {
    if {[botonchan $chan]} {
      # The exact same WHO query eggdrop sends on-join
      putserv "WHO $chan c%chnufat,222"
    }
  }
}
```

vanosg replied:
Do you want this in docs, or can we put this on the wiki? In my head, code samples could/should go there, a section has already been started for this type of thing

thommey commented on doc/sphinx_source/using/accounts.rst:
```diff
@@ -0,0 +1,81 @@
+
+account-tag
+^^^^^^^^^^^
+One supplementary capability that can assist a best-effort account tracking scenario is the IRCv3-defined `account-tag capability <https://ircv3.net/specs/extensions/account-tag>`_. The account-tag capability attaches a tag with the account name associated with the user sending a command. Enabling this capability allows Eggdrop to update its account tracking every time a user talks in channel, se ...
```
Maybe mention `[getaccount $nick $chan]` specifically working in bind callback procs

thommey commented on doc/sphinx_source/using/accounts.rst:
```diff
@@ -0,0 +1,81 @@
+-------------------------------
+The Eggdrop Tcl ACCOUNT bind is triggered whenver an existing account record stored by Eggdrop is modified, such as a user de/authenticating to their account while in a channel, or information such as an account-tag being seen that updates an existing user. However, the ACCOUNT bind will NOT be triggered for the creation of a new user record, such as a user joining a channel. The bind is triggere ...
+
+In order to trigger Tcl script events to cover all instances where a user logs in, you need to pair an ACCOUNT bind with a JOIN bind. This will allow you to execute account-based events when a user joins as well as if they authenticate after joining. 
```
... "If you want to execute an action when Eggdrop first joins a channel and gets to know account names you can trigger on raw 315 (End of /WHO).
```tcl
bind rawt - 315 endofwho
proc endofwho {from key text tags} {
  set chan [lindex [split $text] 0]
  foreach nick [chanlist $chan] {
    if {[getaccount $nick $chan] != "*"} {
      putlog "$nick on $chan is authenticated as [getaccount $nick $chan]"
    } else {
      putlog "$nick on $chan is not authenticated with services"
    }
  }
  # make sure there is no implicit return 1 which would stop further processing
  return 0
}

vanosg replied:
same, doc or wiki?
